src/main/java/bio/terra/service/resourcemanagement/MetadataDataAccessUtils.java

```diff
@@ -89,49 +127,75 @@ public AccessInfoModel accessInfoFromDataset(final Dataset dataset) {
       AzureStorageAccountResource storageAccountResource =
           resourceService.getDatasetStorageAccount(dataset, profileModel);
       return makeAccessInfoAzure(
-          dataset.getName(), storageAccountResource, dataset.getTables(), profileModel);
+          dataset, storageAccountResource, dataset.getTables(), profileModel, userRequest);
     } else {
       throw new IllegalArgumentException("Unrecognized cloud platform");
     }
   }
 
   private AccessInfoModel makeAccessInfoAzure(
-      final String datasetName,
+      final FSContainerInterface collection,
       final AzureStorageAccountResource storageAccountResource,
       final List<? extends Table> tables,
-      final BillingProfileModel profileModel) {
+      final BillingProfileModel profileModel,
+      final AuthenticatedUserRequest userRequest) {
     AccessInfoModel accessInfoModel = new AccessInfoModel();
 
+    BlobSasTokenOptions blobSasTokenOptions =
+        new BlobSasTokenOptions(
+            DEFAULT_SAS_TOKEN_EXPIRATION,
+            new BlobSasPermission().setReadPermission(true).setListPermission(true),
+            userRequest.getEmail());
+
+    String blobName;
+    Function2<FSContainerInterface, Table, String> tableBlobGenerator;
+    if (collection instanceof Dataset) {
```

Comment:
I think where we can infer directly from the object, that seemed easier.  I would have like to use the same thing in the other class but the problem is that the contexts where it gets called don't always have access to the full FSContainerInterface object

Reply:
If you don't want to add another argument, a better solution IMO would be to define an `interface` that returns `CollectionType` (or another enum type) and have `Dataset` and `Snapshot` implement it, so our code doesn't need to use `instanceof`. Or you could add a method to the interface that declares this `BiFunction` and then you can call the function on `collection ` and remove the `if`/`else`.

Reply:
I was thinking that but it's actually worse from a performance standpoint.  That forces us to do a lookup for the database when we only have the id and I want to avoid having to do that in all cases

Reply:
If we have a `Dataset` (or `Snapshot`) which implements an interface API `getCollectionType()` then there's no database call necessary, since these objects know which kind of collection they are. Or are you saying you don't want to have to use a `collection` object at all here?

Reply:
Datasets and Snapshots are created by fetching from the DB (which is actually several calls).  I'm trying to avoid having to do that.  We could build them and just set the ID but then you end up in a space where you have to know what fields are in the dataset...it feels hacky

Reply:
I still don't get it. This code is called at the point when you've already made a DB call and the result (`collection`) is examined to determine if you have a dataset or a snapshot. You don't need to make another DB call since you already have the object. If you add `getContainerType()` to `FSContainerInterface` then you can avoid the `instanceof`.

Reply:
Oh sorry, I was thinking about the other code that uses that enum.  Sure, I can make snapshot and dataset return the values of the enum in a shared method and use that instead of instanceof

Reply:
(BTW, I did that in a separate commit)